src/Scale.hs

```diff
@@ -1,99 +1,199 @@
 module Scale
   ( Scale(..)
-  , scaleToIntervals
+  , BaseMode(..)
+  , Mode(..)
+  , ScaleExt(..)
+  , baseModeIntervals
+  , modeToIntervals
+  , scaleToNotes
+  , modalDistance
+  , modesToExts
+  , intervalsToMode
+  , zipToIntervalSet
+  , isSubsetMode
+  , getSubsetModeByDegree
+  , invert
   ) where
 
+
 import Base.Core.Quality.IQuality
+import Base.Chord.Root
+import Base.Interval hiding (invert)
+import qualified Base.Interval as I (invert)
+import Base.Core.Accidental(Accidental(..), impliedShift, shiftToAcc, natural)
+import Data.List (sort, sortBy, intercalate, takeWhile)
+import Data.Set(Set(..), fromList, toAscList, elemAt, insert, delete, mapMonotonic, isSubsetOf)
+import qualified Data.Set as S(filter, map)
+import Data.Maybe(fromJust)
+import Data.Function
+
+
+
+data Scale = Scale Root Mode
+
+instance Show Scale where
+  show (Scale root mode) = show root ++ show mode
+
+data Mode = Mode BaseMode [ScaleExt]
+
+instance Show Mode where
+  show (Mode base exts) = show base
+                       --Add a space if there are extensions...
+                       ++ (replicate (signum (length (exts))) ' ')
+                       --Add extensions separated by a comma...
+                       ++ (intercalate ", " (show <$> exts))        
 
-import Base.Interval (Interval, intervalFrom, (|+|), (|-|))
-import Data.List (sort)
-
-
-data Scale
-  = SLydian
-  | SDorian
-  | SMixolydian
-  | SAugmentedQuality
-  | SDiminishedQuality
-  | SMajor
-  | SMinor
-  | SPhrygian
-  | SLocrian
-  | SMelodicMinor
-  | SDorianb2
-  | SLydianAug
-  | SLydianDom
-  | SMixolydianb6
-  | SLocrianSharp2
-  | SAltered
-  | SHarmonicMinor
-  | SLocrianNat6
-  | SIonianSharp5
-  | SDorianSharp4
-  | SPhrygianDom
-  | SLydianSharp2
-  | SSuperLocrianbb7
-
-
-nthDegreeIntervals :: [Interval] -> Int -> [Interval]
-nthDegreeIntervals ints n = sort $ (|-| rootInterval) <$> ints
+data ScaleExt = ScaleExt { acc :: Accidental
+                         , deg :: Int
+                         }
```

Comment:
I was wondering if the ScaleExt is the general, more abstract idea of extensions, and our version in Base.Chord.Extension is a special case where there's max 1 sharp/flat. I don't think reconciling the 2 would make it easier to parse, but it might let us consolidate some repetitive code?

Reply:
Yep! That's exactly what I was saying: `ScaleExt` carries more information than `Extension`, and the information carried by `Extension` is a strict subset, so there's no reason to have both types. `Base.Chord.Extension` can be updated to contain the number of accidentals, which will likely require changes in lots of other places (the type checker should help.)